pilot/pkg/serviceregistry/kube/controller/remotekubeconfigcontroller.go

```diff
@@ -0,0 +1,187 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"istio.io/istio/pkg/kube"
+	"istio.io/pkg/log"
+)
+
+const (
+	initialSyncSignal = "INIT"
+	maxRetries        = 5
+)
+
+// RemoteKubeConfigController is the controller implementation for `istio-kubeconfig` secret
+type RemoteKubeConfigController struct {
+	kubeclientset kubernetes.Interface
+	queue         workqueue.RateLimitingInterface
+	informer      cache.SharedIndexInformer
+	mu            sync.RWMutex
+	initialSync   bool
+}
+
+// NewRemoteKubeConfigController returns a pointer to a newly constructed RemoteKubeConfigController instance.
+func NewRemoteKubeConfigController(kubeClient kubernetes.Interface, ns, name string) *RemoteKubeConfigController {
+	secretsInformer := cache.NewSharedIndexInformer(
+		&cache.ListWatch{
+			ListFunc: func(opts metav1.ListOptions) (runtime.Object, error) {
+				opts.FieldSelector = "metadata.name=" + name
+				return kubeClient.CoreV1().Secrets(ns).List(context.TODO(), opts)
+			},
+			WatchFunc: func(opts metav1.ListOptions) (watch.Interface, error) {
+				opts.FieldSelector = "metadata.name=" + name
+				return kubeClient.CoreV1().Secrets(ns).Watch(context.TODO(), opts)
+			},
+		},
+		&corev1.Secret{}, 0, cache.Indexers{},
+	)
+	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+	controller := &RemoteKubeConfigController{
+		kubeclientset: kubeClient,
+		informer:      secretsInformer,
+		queue:         queue,
+	}
+	log.Info("Setting up event handlers")
+	secretsInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, _ := cache.MetaNamespaceKeyFunc(obj)
+			log.Infof("secret % add: %s", key)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			if oldObj == newObj || reflect.DeepEqual(oldObj, newObj) {
+				return
+			}
+
+			key, err := cache.MetaNamespaceKeyFunc(newObj)
+			log.Infof("Processing update: %s", key)
+			if err == nil {
+				queue.Add(key)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			log.Infof("Processing delete: %s", key)
+			if err == nil {
+				queue.Add(key)
+			}
```

Comment:
We donot need to process add and delete

Reply:
For add, I just log it out. For delete, processed it to warn user that the secret has been deleted incorrectly.

Reply:
For delete just warn it

Reply:
And no need to enqueue it